File: javascript/6kyu/breakCamelCase.js
//Reflection: Regex keeps getting cooler. When using replace, () allows you to capture an expression. $1 allows you to reference that expression


//Problem:
// Complete the solution so that the function will break up camel casing, using a space between words.

// Example
// solution('camelCasing') // => should return 'camel Casing'


//Intial correct solution:
function solution(string) {
  let words = [];
  let lastIndex = 0;
  string.split('').forEach(function(char, idx) {
    if (/[A-Z]/.test(char)) {
      words.push(string.slice(lastIndex, idx));
      lastIndex = idx;
    }
  })
  words.push(string.slice(lastIndex));
  return words.join(' ');
}

//Interesting solutions
function solution(string) {
	return string.replace(/([a-z])([A-Z])/g, '$1 $2');
}

File: ruby/6kyu/break_camel_case.rb
#Reflection: ruby uses \1 instead of $1 to capture the regex match. 

#Problem:
# Complete the solution so that the function will break up camel casing, using a space between words.

# Example
# solution('camelCasing') # => should return 'camel Casing'

#inital correct solution
def solution(string)
    string.gsub(/([a-z]+)([A-Z]+)/, '\1 \2')
end

#other creative solutions
def solution(string)
    string.gsub(/([A-Z])/, ' \1')
end

def solution(string)
    string.gsub(/(?=[A-Z])/, ' ') # ?= is a positive lookahead
end